client/src/main/scala/com.crobox.clickhouse/internal/ClickHouseExecutor.scala

```diff
@@ -149,6 +224,132 @@ object ClickHouseExecutor {
 
   }
 
+  /**
+   * Clickhouse send http progress headers with the name X-ClickHouse-Progress which cannot be handled in a streaming matter in akka
+   * In the request we include our own custom header `X-Internal-Identifier` so we can send the internal query id with the progress
+   * We expect to first find the request with the custom header and then receive the progress headers
+   *  The progress headers are being intercepted by the transport and send to the source as progress events with the internal query id
+   * After the first progress header is received we send the end of headers character so that akka can eagerly return the http response
+   * */
+  class ProgressHeadersAsBodyClientTransport(source: SourceQueue[String]) extends ClientTransport {
+    override def connectTo(
+        host: String,
+        port: Int,
+        settings: ClientConnectionSettings
+    )(implicit system: ActorSystem): Flow[
+      ByteString,
+      ByteString,
+      Future[Http.OutgoingConnection]
+    ] =
+      BidiFlow
+        .fromGraph(new ProgressHeadersAsEventsStage(source))
+        .joinMat(
+          ClientTransport.TCP
+            .connectTo(host, port, settings)
+        )((_, result) => result)
+  }
+
+  class ProgressHeadersAsEventsStage(source: SourceQueue[String])
+      extends GraphStage[BidiShape[ByteString, ByteString, ByteString, ByteString]] {
+
+    val ClickhouseProgressHeader = "X-ClickHouse-Progress"
+
+    val in1                = Inlet[ByteString]("ProgressHeadersAsEvents.in1")
+    val out1               = Outlet[ByteString]("ProgressHeadersAsEvents.out1")
+    val in2                = Inlet[ByteString]("ProgressHeadersAsEvents.in2")
+    val out2               = Outlet[ByteString]("ProgressHeadersAsEvents.out2")
+    val Crlf               = "\r\n"
+    val LookAheadCRLFSplit = "(?<=\\r\\n)"
+
+    override val shape = BidiShape.of(in1, out1, in2, out2)
+    override def createLogic(
+        inheritedAttributes: Attributes
+    ): GraphStageLogic = new GraphStageLogic(shape) with StageLogging {
+      var queryId: Option[String] = None
+      var queryIsInProgress       = false
+      setHandler(
+        in1,
+        new InHandler {
+          override def onPush(): Unit = {
+            val byteString = grab(in1)
+            if (byteString.containsSlice(ByteString(InternalQueryIdentifier))) {
+              val incomingString  = byteString.utf8String
+              val responseStrings = incomingString.split(Crlf)
+              val queryIdHeader   = responseStrings.find(_.contains(InternalQueryIdentifier))
+              if (queryIdHeader.isEmpty) {
+                log.warning(s"Could not extract the query id from the containing $incomingString")
+              }
+              if (queryIsInProgress) {
+                log.warning("The previous query was not terminated correctly")
```

Comment:
error?

Reply:
Same as before, it's not a terminal error.

Reply:
it could be? errors on the side of the database could lead to inconsistency / dataloss very quickly.

Reply:
This does not denote an error on the database level, only an error in the way we handle the progress. The query will still be executed completely